src/Data/Formatter/DateTime.purs

```diff
@@ -0,0 +1,426 @@
+  Milliseconds a → "SSS" <> cb a
+  Placeholder s a → s <> cb a
+  End → ""
+
+printFormatter ∷ Formatter → String
+printFormatter f = printFormatterF printFormatter $ unroll f
+
+parseFormatString ∷ String → Either String Formatter
+parseFormatString s =
+  lmap (\(P.ParseError {message}) → message) $ P.runParser s formatParser
+
+
+placeholderContent ∷ P.Parser String String
+placeholderContent =
+  map Str.fromCharArray
+    $ PC.try
+    $ Arr.some
+    $ PS.noneOf
+    $ Str.toCharArray "YQMDXWEHhamsS"
+
+formatterFParser
+  ∷ ∀ a
+  . P.Parser String a
+  → P.Parser String (FormatterF a)
+formatterFParser cb =
+  PC.choice
+    [ (PC.try $ PS.string "YYYY") *> map YearFull cb
+    , (PC.try $ PS.string "YY") *> map YearTwoDigits cb
+    , (PC.try $ PS.string "Y") *> map YearAbsolute cb
+    , (PC.try $ PS.string "MMMM") *> map MonthFull cb
+    , (PC.try $ PS.string "MMM") *> map MonthShort cb
+    , (PC.try $ PS.string "MM") *> map MonthTwoDigits cb
+    , (PC.try $ PS.string "DD") *> map DayOfMonth cb
+    , (PC.try $ PS.string "E") *> map DayOfWeek cb
+    , (PC.try $ PS.string "HH") *> map Hours24 cb
+    , (PC.try $ PS.string "hh") *> map Hours12 cb
+    , (PC.try $ PS.string "a") *> map Meridiem cb
+    , (PC.try $ PS.string "mm") *> map Minutes cb
+    , (PC.try $ PS.string "ss") *> map Seconds cb
+    , (PC.try $ PS.string "SSS") *> map Milliseconds cb
+    , (Placeholder <$> placeholderContent <*> cb)
+    , (PS.eof $> End)
+    ]
+
+formatParser ∷ P.Parser String Formatter
+formatParser =
+  Lazy.fix \f → map roll $ formatterFParser f
+
+formatF
+  ∷ ∀ a
+  . (a → String)
+  → DT.DateTime
+  → FormatterF a
+  → String
+formatF cb dt@(DT.DateTime d t) = case _ of
+  YearFull a →
+    (show $ fromEnum $ D.year d) <> cb a
+  YearTwoDigits a →
+    show ((fromEnum $ D.year d) - 1900) <> cb a
+  YearAbsolute a →
+    show (fromEnum $ D.year d) <> cb a
+  MonthFull a →
+    show (D.month d) <> cb a
+  MonthShort a →
+    printShortMonth (D.month d) <> cb a
+  MonthTwoDigits a →
+    show (fromEnum $ D.month d) <> cb a
+  DayOfMonth a →
+    show (fromEnum $ D.day d) <> cb a
+  UnixTimestamp a →
+    (show $ Int.floor $ (_ / 1000.0) $ Dur.unMilliseconds $ unInstant $ fromDateTime dt) <> cb a
+  DayOfWeek a →
+    show (fromEnum $ D.weekday d) <> cb a
+  Hours24 a →
+    show (fromEnum $ T.hour t) <> cb a
+  Hours12 a →
+    show ((fromEnum $ T.hour t) `mod` 12) <> cb a
+  Meridiem a →
+    (if (fromEnum $ T.hour t) > 12 then "PM" else "AM") <> cb a
+  Minutes a →
+    show (fromEnum $ T.minute t) <> cb a
+  Seconds a →
+    show (fromEnum $ T.second t) <> cb a
+  Milliseconds a →
+    show (fromEnum $ T.millisecond t) <> cb a
+  Placeholder s a →
+    s <> cb a
+  End → ""
+
+
+format ∷ Formatter → DT.DateTime → String
+format f dt = formatF (flip format dt) dt $ unroll f
+
+formatDateTime ∷ String → DT.DateTime → Either String String
+formatDateTime pattern datetime =
+  parseFormatString pattern <#> flip format datetime

```

Comment:
I kinda have a problem with this name, as both functions do indeed "format date/time" so it seems odd to prefix one with that. I don't really have a suggestion for a better name though, aside from my old fallback to `format'`...